### DISCUSSION TITLE
When using TemplateProcessor for serveStatic pages do not respond "304 Not Modified"

### DISCUSSION THREAD

**Discussion by mawi42** [mathieucarbou/ESPAsyncWebServer#147] [General]
When template processor is used do not return "not modifed 304" but refresh the webpage. Template processors might be used for non static pages since some variables which will be replaced might change.

Proposed workaround instead of introducing new method serveDynamic, use info if templat processor is set (_callback).
 
Webhandlers.cpp line 218:
```
if (!_callback && _last_modified.length() && _last_modified == request->header(T_IMS)) {
      request->_tempFile.close();
      request->send(304); // Not modified
    } else if(!_callback && _cache_control.length() && request->hasHeader(T_INM) && request->header(T_INM).equals(etag)) {

```

**Comment by mathieucarbou** [comment]
Hello,

Your proposal cannot be implemented like you suggest, otherwise this is a behaviour change for any person who are using the callback mechanism and still have content that can be cached.

Could you please elaborate more on your use case with a code sample instead of suggesting a fix which breaks existing behaviour ?

**`AsyncStaticWebHandler` is a subclass specifically designed to serve static content that can be cached.**

If this is not what you want, then, don't use it...

Thanks :-)

**Reply by mawi42** [reply]
OK I understand, sorry for bad explanation. I was not sure  which design is preferred, just adding a paramter (like "no caching"), new kind of class "serveDynamic". But than mabe new base class is needed and inherit 2 for static and dynamic content, which results maybe in a too big change. ..... 
I understood my proposal not to be the fix but to show what kind of change is needed to allow this functionality.

The problem is "just" that a variable added by the template processor is not static but changes its value. So repsonding 304 is not correct in this case. But all other functionality (beside the caching) can be the same.

**Reply by mawi42** [reply]
I was thinking a bit more on this topic, and my proposal would be to add a member variable like` _response304 = true` (default) to the `AsyncStaticWebHandler ` class, which can be disabled via a new method `disableResponse304`. The member variable could be used to disable the 304 respinse instead of my original proposal using _callback state. I think this does not break anything, allows the feature and the change / effort is minimal. 
I tried to make a PR but seems, I am not allowed to do. I also could provide a simple sample, actually I just add into template processor a counter which illustrates this well.

**Reply by mathieucarbou** [reply]
Like I said, instead of proposing a fix to a class I won't fix, please show a .ino sample of what you want to achieve and why.

Like i said, `AsyncStaticWebHandler`'s goal is to serve STATIC content that NEVER changes. If you use a template with it, the template should always generate the same output.

The code you showed me aims at updating this class, which I will always refuse to do the way you propose because the content will be allowed to change over.

The templating feature is also accessible from other methods. I think you should use te other ones that let you control the caching and content the way you want. 

But stop proposing a fix to a class without first having showed a complete use case of what you want to achieve ;-)

**Reply by mawi42** [reply]
Hi again, maybe you got me wrong, I understand that you do not want to break anything, I also understand that the current name of the class indicates static pages. I just want to avoid to big changes. Most likely you can get the same using existing handler classes, but I feel using it a simple way is more user firendly.
Attached is the ino, but I have to say that just the simple sample surprisingly works - which is strange - so I guess from your side that means you can close this issue and and I'm sorry for bothering you and the missunderstanding. 
However I will investigate why in my combination it is not working (no counting) and I receive 304, where now the sample ino reloads the page and shows allways a new number.
[Dynamic.ino.txt](https://github.com/user-attachments/files/17552782/Dynamic.ino.txt)

**Reply by mathieucarbou** [reply]
I've looked at your example and I already told you that the library already does what you need in terms of API.

servestatic is used to serve static content. Point. So please do not use it and look for the other available methods 

What you want is to serve a file (template) with a template processor.

Have a look at what's available:

https://github.com/mathieucarbou/ESPAsyncWebServer/blob/main/src/ESPAsyncWebServer.h#L323

There is no need to have a seve dynamic method.

All the current methods allow you to support any use case since you have control over everything. Servestatic for example is a simplified version of file serving + cache headers.

Of you see a use case hat is not possible by creating your own response handler or a use case where you seem to do the same code in all handlers (maybe a common use case,) then we will see.

But instead of patching something that cannot be modified like you want , please first try to do what you want with the existing methods and you will surely see that it is possible and you do not need to modify servestatic.

**Comment by mathieucarbou** [comment]
Please have a look at the examples that I have just updated to add to cover your use case:

https://github.com/mathieucarbou/ESPAsyncWebServer/blob/main/examples/SimpleServer/SimpleServer.ino#L415-L436

**Reply by mawi42** [reply]
OK, I see, thank you for pointing me to this, I missed your hint that other methods cover this
